playbooks/bootstrap-sudo.yml

```diff
@@ -0,0 +1,35 @@
+---
+- hosts: all
+  tasks:
+    - name: check for presence of sudo
+      command: which sudo
+      changed_when: false
+      failed_when: false
+      register: sudo
+
+    # If sudo is installed, try `become_method=sudo`
+    - name: use method "sudo" (requires password for "{{ ansible_user }}")
+      set_fact: method=sudo
+      when: sudo.rc == 0
+
+    # Otherwise, use `become_method=su`
+    - name: use method "su" (requires password for "root")
+      set_fact: method=su
+      when: sudo.rc != 0
+
+    - name: install sudo package
+      package: name=sudo
+      become: true
+      become_method: "{{ method }}"
+
+    - name: "grant password-less sudo to user \"{{ ansible_user }}\""
+      copy:
+        content: "{{ ansible_user }} ALL=(ALL) NOPASSWD:ALL"
```

Comment:
I think this var needs to be `{{ ansible_user | default(ansible_env.SUDO_USER) }}`

Reply:
I saw you keep using that, but it never seemed necessary for me. Can you give me a repro case where that's useful?

Reply:
If I pass `-b` to `ansible-playbook`, `{{ ansible_user }}` is set to `root`. In this case it would grant password-less sudo to `root`.

Reply:
I want to believe you, but I just can't find a way to replicate that.
```
$ ansible localhost -m debug -a 'var=ansible_user'
localhost | SUCCESS => {
    "ansible_user": "vagrant"
}
$ ansible localhost -b -m debug -a 'var=ansible_user'
localhost | SUCCESS => {
    "ansible_user": "vagrant"
}
```
(tried with other hosts besides localhost, too)

Reply:
You are defining `ansible_user`, so the variable exists. If I have the same user on the control and target machines, for example, meaning I don't need to set that variable by hand, then it does not work:

```$ ansible lab-gk8-worker1 -m debug -a 'var=ansible_user'
lab-gk8-worker1 | SUCCESS => {
    "ansible_user": "VARIABLE IS NOT DEFINED!"
}
```

I don't remember the case where it comes back as `root`, the conversation is buried in slack/github somewhere. Maybe it was a different situation. Regardless, my suggestion always returns the correct user without requiring someone to set `{{ ansible_user }}`.

Reply:
That was it, good call. I was setting the user in the inventory.

However, your solution is still not good. Check this out:

**ansible_env depends on whether `setup` was run with --become or not**
```bash
# run setup as root
$ ansible virtual-login -b -m setup

# ... and SUDO_USER is SET with or without --become
$ ansible virtual-login -m debug -a "msg='<{{ansible_env.USER|default(\"\")}}> <{{ansible_env.SUDO_USER|default(\"\")}}>'"
virtual-login | SUCCESS => {
    "msg": "<root> <vagrant>"
}
$ ansible virtual-login -b -m debug -a "msg='<{{ansible_env.USER|default(\"\")}}> <{{ansible_env.SUDO_USER|default(\"\")}}>'"
virtual-login | SUCCESS => {
    "msg": "<root> <vagrant>"
}

# run setup without root
$ ansible virtual-login -m setup

# ... and SUDO_USER is UNSET with or without --become
$ ansible virtual-login -m debug -a "msg='<{{ansible_env.USER|default(\"\")}}> <{{ansible_env.SUDO_USER|default(\"\")}}>'"
virtual-login | SUCCESS => {
    "msg": "<vagrant> <>"
}
$ ansible virtual-login -b -m debug -a "msg='<{{ansible_env.USER|default(\"\")}}> <{{ansible_env.SUDO_USER|default(\"\")}}>'"
virtual-login | SUCCESS => {
    "msg": "<vagrant> <>"
}

```
**`shell`, for reference, doesn't depend on `setup`**
```bash
$ ansible virtual-login -m shell -a 'echo "<$USER> <$SUDO_USER>"'
virtual-login | CHANGED | rc=0 >>
<vagrant> <>

$ ansible virtual-login -b -m shell -a 'echo "<$USER> <$SUDO_USER>"'
virtual-login | CHANGED | rc=0 >>
<root> <vagrant>
```

So ...

1. We can't rely on `ansible_user` because it might not be set
2. We can't rely on `SUDO_USER` because it's only set if `setup` was run as root
3. We can't really rely on `ansible_env` at all because we have fact caching turned on

I think this is the only safe thing to do:
```yaml
    # re-run setup, to guarantee the cache is good
    - setup: gather_subset=min
    # use SUDO_USER if set, otherwise USER
    - debug: msg={{ansible_env.SUDO_USER|default(ansible_env.USER)}}
```